Stop capturing raw pointers in RunLoop::Timer constructors

cdumez opened the pull request:
#### 9f9402e226ed168712b8cafddb7c96b9b53354f2
<pre>
Stop capturing raw pointers in RunLoop::Timer constructors
<a href="https://bugs.webkit.org/show_bug.cgi?id=301496">https://bugs.webkit.org/show_bug.cgi?id=301496</a>

Reviewed by Ryosuke Niwa.

Stop capturing raw pointers in RunLoop::Timer constructors, as this isn&apos;t safe.
Capture weak pointers instead.

* Source/JavaScriptCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/JavaScriptCore/runtime/JSRunLoopTimer.cpp:
(JSC::JSRunLoopTimer::Manager::singleton):
* Source/JavaScriptCore/runtime/JSRunLoopTimer.h:
(JSC::JSRunLoopTimer::Manager::ref const): Deleted.
(JSC::JSRunLoopTimer::Manager::deref const): Deleted.
* Source/WTF/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/linux/RealTimeThreads.h:
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.h:
* Source/WebCore/platform/MainThreadSharedTimer.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepadProvider.h:
* Source/WebCore/platform/gamepad/libwpe/GamepadProviderLibWPE.h:
* Source/WebCore/platform/gamepad/manette/ManetteGamepadProvider.h:
* Source/WebCore/platform/graphics/cg/IOSurfacePool.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.h:
* Source/WebCore/platform/graphics/win/DisplayRefreshMonitorWin.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.h:
* Source/WebKit/NetworkProcess/glib/DNSCache.h:
* Source/WebKit/Shared/API/APISerializedScriptValue.cpp:
* Source/WebKit/UIProcess/API/glib/IconDatabase.h:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.h:
* Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.h:
* Source/WebKit/UIProcess/PerActivityStateCPUUsageSampler.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/ResponsivenessTimer.h:
* Source/WebKit/UIProcess/WebBackForwardCacheEntry.h:
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/ios/GestureRecognizerConsistencyEnforcer.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedDisplayRefreshMonitorPlayStation.h:
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:

Canonical link: <a href="https://commits.webkit.org/302261@main">https://commits.webkit.org/302261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6e15413ca44feee5813fc670845e8ad1974bd35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/732 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39294 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135857 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79908 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d6e9f0d3-7338-430e-88bf-560bae49408d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130335 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/605 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97794 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65702 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3d05c8dc-a934-49dd-ba0b-e2092dca97e2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131411 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/489 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115095 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78405 "Found 1 new API test failure: WPE/TestWebsiteData:/webkit/WebKitWebsiteData/itp (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2b06501f-45ed-4285-8727-908152172043) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33203 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79139 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/120477 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108856 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33686 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138306 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/126923 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/570 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/540 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106335 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/613 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111436 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106145 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27053 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/476 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29977 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52905 "Hash f6e15413 for PR 53018 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/626 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63820 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/159946 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/516 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39947 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/579 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/583 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/platform/gamepad/libwpe/GamepadProviderLibWPE.h:
```diff
@@ -39,13 +40,11 @@ namespace WebCore {
+class GamepadProviderLibWPE final : public GamepadProvider, public CanMakeWeakPtr<GamepadProviderLibWPE> {
     WTF_MAKE_NONCOPYABLE(GamepadProviderLibWPE);
     friend class NeverDestroyed<GamepadProviderLibWPE>;
-
```
I think this blank line is useful to separate member functions.

rniwa commented on Source/WebCore/platform/gamepad/libwpe/GamepadProviderLibWPE.h:
```diff
@@ -39,13 +40,11 @@ namespace WebCore {
-
 public:
     static GamepadProviderLibWPE& singleton();
-
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSRunLoopTimer.cpp:
```diff
@@ -110,12 +111,13 @@ void JSRunLoopTimer::Manager::timerDidFire()
-    static Manager* manager;
+    static LazyNeverDestroyed<std::unique_ptr<Manager>> manager;
     static std::once_flag once;
     std::call_once(once, [&] {
-        manager = new Manager;
+        auto newManager = std::unique_ptr<Manager> { new Manager };
+        manager.construct(WTFMove(newManager));
     });
-    return *manager;
+    return *manager.get();
```
How about making it `LazyNeverDestroyed<Manager>`? If this is singleton, I don't think we need to allocate memory from heap.

cdumez committed:
Stop capturing raw pointers in RunLoop::Timer constructors
https://bugs.webkit.org/show_bug.cgi?id=301496

Reviewed by Ryosuke Niwa.

Stop capturing raw pointers in RunLoop::Timer constructors, as this isn't safe.
Capture weak pointers instead.

* Source/JavaScriptCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/JavaScriptCore/runtime/JSRunLoopTimer.cpp:
(JSC::JSRunLoopTimer::Manager::singleton):
* Source/JavaScriptCore/runtime/JSRunLoopTimer.h:
(JSC::JSRunLoopTimer::Manager::ref const): Deleted.
(JSC::JSRunLoopTimer::Manager::deref const): Deleted.
* Source/WTF/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/linux/RealTimeThreads.h:
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.h:
* Source/WebCore/platform/MainThreadSharedTimer.h:
* Source/WebCore/platform/gamepad/cocoa/GameControllerGamepadProvider.h:
* Source/WebCore/platform/gamepad/libwpe/GamepadProviderLibWPE.h:
* Source/WebCore/platform/gamepad/manette/ManetteGamepadProvider.h:
* Source/WebCore/platform/graphics/cg/IOSurfacePool.h:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.h:
* Source/WebCore/platform/graphics/win/DisplayRefreshMonitorWin.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebCore/platform/mock/MockAudioDestinationCocoa.h:
* Source/WebKit/NetworkProcess/glib/DNSCache.h:
* Source/WebKit/Shared/API/APISerializedScriptValue.cpp:
* Source/WebKit/UIProcess/API/glib/IconDatabase.h:
* Source/WebKit/UIProcess/Automation/SimulatedInputDispatcher.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionAlarm.h:
* Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.h:
* Source/WebKit/UIProcess/PerActivityStateCPUUsageSampler.h:
* Source/WebKit/UIProcess/ProcessThrottler.cpp:
* Source/WebKit/UIProcess/ProcessThrottler.h:
* Source/WebKit/UIProcess/ResponsivenessTimer.h:
* Source/WebKit/UIProcess/WebBackForwardCacheEntry.h:
* Source/WebKit/UIProcess/WebProcessCache.h:
* Source/WebKit/UIProcess/ios/GestureRecognizerConsistencyEnforcer.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/AcceleratedSurface.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedDisplayRefreshMonitorPlayStation.h:
* Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:

Canonical link: https://commits.webkit.org/302261@main
